src/makerom/rom_header.h

```diff
@@ -0,0 +1,143 @@
+ */
+#define PI_DOMAIN_1_CFG(lat, pwd, pgs, rls)   \
+    .byte (((rls) & 3) << 4) | ((pgs) & 0xF); \
+    .byte (pwd) & 0xFF;                       \
+    .byte (lat) & 0xFF
+
+/**
+ * Some older libultra versions use this field to set osClockRate. It does not have any other meaningful effect, the
+ * clock rate of physical units does not actually change to match this value.
+ */
+#define SYSTEM_CLOCK_RATE_SETTING(num) \
+    .word (num)
+
+/**
+ * Indicates the entrypoint address of the program that IPL3 will load the boot segment to and execute.
+ * This should be >= 0x80000400.
+ */
+#define ENTRYPOINT(sym) \
+    .word (sym)
+
+/**
+ * Indicates the hardware revision the program is designed for (hw_major, hw_minor) and what libultra version (os_ver)
+ * it uses.
+ * The hardware revision for a retail N64 is (2,0).
+ * The libultra version may be a single letter, without quotes.
+ */
+#define LIBULTRA_VERSION(hw_major, hw_minor, os_ver)   \
+    .half 0;                                           \
+    .byte (hw_major) * 10 + (hw_minor);                \
+    _os_ver_start = .;                                 \
+    .ascii #os_ver ;                                   \
+    .if (. - _os_ver_start) != 1;                      \
+        .error "OS version should be just one letter"; \
+    .endif
+
+/**
+ * ROM Checksum value. IPL3 computes a checksum over ROM data in the range 0x1000 to 0x101000 and compares it to this
+ * value, if the results differ it will refuse to boot the program.
+ *
+ * This can generally be left as 0 in the ROM header declaration and filled in after the full ROM image is available to
+ * compute the checksum with.
+ */
+#define CHECKSUM(value)                 \
+    .word ((value) >> 32) & 0xFFFFFFFF; \
+    .word ((value) >>  0) & 0xFFFFFFFF
+
+/**
+ * For unused header space. Fills num bytes with 0.
+ */
+#define PADDING(num) \
+    .fill (num)
+
+/**
+ * Defines the ROM name. This should be a string that is at most 20 characters long, a null terminator is not required.
+ * If a name is less than 20 characters, the remaining space will be padded with spaces.
+ */
+#define ROM_NAME(name)                                             \
+    _name_start = .;                                               \
+    .ascii name;                                                   \
+    .if (. - _name_start) > 20;                                    \
+        .error "ROM name too long, must be at most 20 characters"; \
+    .endif;                                                        \
+    .if (. - _name_start) < 20;                                    \
+        .fill 20 - (. - _name_start), 1, 0x20;                     \
+    .endif
+
+/**
+ * Identifies the storage medium the program intends to use.
+ * Should be one of:
+ *  - CARTRIDGE
+ *  - CARTRIDGE_EXPANDABLE
+ *  - DISK
+ *  - DISK_EXPANSION
+ */
+#define MEDIUM(type) \
+    .ascii STORAGE_MEDIUM_##type
```

Comment:
IMO token pasting should be avoided but I see the appeal of not being redundant here. What about writing something like `CHAR(STORAGE_MEDIUM_CARTRIDGE)` instead?

Reply:
I prefer token pasting since it restricts the inputs to the macro to be one of the valid options, and it reads better in the asm file. I don't really see an issue with token pasting in general so this may need more opinions.

Reply:
I generally dont like pasting from a discovery standpoint, when searching globally for enum values. but I dont have an issue with isolated use cases like this and think this is fine

Reply:
> restricts the inputs to the macro to be one of the valid options

This seems reason enough for me to do token pasting.

Reply:
I also think token pasting makes it harder to search for things in general (for example I have been confused by `PRIMITIVE` and `TEXEL0` from the gbi macros), but it's not a big deal in this case so feel free to ignore